tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs

```diff
@@ -0,0 +1,168 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.IO.Pipelines.Tests
+{
+    public class PipeToStreamFacts
+    {
+        [Fact]
+        public async Task CopyFromPipeToEnd()
+        {
+            var buffer = new byte[1000];
+            (new Random()).NextBytes(buffer);
+            using (var pipeFactory = new PipeFactory())
+            using (var stream = new MemoryStream())
+            {
+                var pipe = pipeFactory.Create();
+                await pipe.Writer.WriteAsync(buffer);
+                pipe.Writer.Complete();
+                await pipe.Reader.CopyToEndAsync(stream);
+
+                Assert.Throws<InvalidOperationException>(() =>
+                {
+                    var reader = pipe.Reader.ReadAsync();
+                });
+                Assert.Equal(buffer, stream.ToArray());
+            }
+        }
+
+        [Fact] 
+        public async Task CopyFromPipeToEndObserveStreamThrow()
+        {
+            var buffer = new byte[1000];
+            (new Random()).NextBytes(buffer);
+            using (var pipeFactory = new PipeFactory())
+            using (var stream = new FakeExceptionStream())
+            {
+                var pipe = pipeFactory.Create();
+                await pipe.Writer.WriteAsync(buffer);
+                pipe.Writer.Complete();
+                await pipe.Reader.CopyToEndAsync(stream);
+                await Assert.ThrowsAsync<InvalidOperationException>(async () =>
+                {
+                    var reader = await pipe.Reader.ReadAsync();
+                });
+            }
+        }
+
+        [Fact]
+        public async Task CopyFromPipeObserveStreamThrow()
+        {
+            var buffer = new byte[1000];
+            (new Random()).NextBytes(buffer);
+            using (var pipeFactory = new PipeFactory())
+            using (var stream = new FakeExceptionStream())
+            {
+                var pipe = pipeFactory.Create();
+                await pipe.Writer.WriteAsync(buffer);
+                pipe.Writer.Complete();
+
+                await Assert.ThrowsAsync<IOException>(async () =>
+                {
+                    await pipe.Reader.CopyToAsync(stream);
+                });
+            }
+        }
+
+        [Fact]
+        public async Task CopyFromPipeNoEnd()
+        {
+            var buffer = new byte[1000];
+            (new Random()).NextBytes(buffer);
+            using (var pipeFactory = new PipeFactory())
+            using (var stream = new MemoryStream())
+            {
+                var pipe = pipeFactory.Create();
+                await pipe.Writer.WriteAsync(buffer);
+                await pipe.Reader.CopyToAsync(stream);
+                var reader = pipe.Reader.ReadAsync();
+
+                Assert.False(reader.IsCompleted);
+                Assert.Equal<byte>(buffer, stream.ToArray());
+            }
+        }
+
+        [Fact]
+        public async Task CopyFromStreamToEnd()
```

Comment:
Like the first test, this could also use a better name. Maybe CopyFromStreamToEndCompletesPipeWriter.